service/storage_proxy.cc

```diff
@@ -5260,6 +5271,121 @@ const db::hints::host_filter& storage_proxy::get_hints_host_filter() const {
+        ).then([&sp, id, guard = sp.shared_from_this()] (std::tuple<future<>, future<>>&& f_tup) {
+            auto handle_result = [id] (future<> f, sstring kind) {
+                if (!f.failed()) {
+                    slogger.debug("Hint sync point {} for {} reached", id, kind);
+                    f.get();
+                } else {
+                    slogger.debug("An error occured when waiting for hint sync point {} for {} to resolve: {}", id, kind, f.get_exception());
+                }
+            };
+            handle_result(std::move(std::get<0>(f_tup)), "hints manager");
+            handle_result(std::move(std::get<1>(f_tup)), "hints for views manager");
+            sp._hint_queue_checkpoints.erase(id);
+        });
+
+        return make_ready_future<utils::UUID>(id);
+    });
+}
+
+future<bool> storage_proxy::check_hint_queue_sync_point(utils::UUID sync_point) {
+    return container().invoke_on(0, [sync_point] (storage_proxy& sp) {
+        return !sp._hint_queue_checkpoints.contains(sync_point);
+    });
+}
+
+future<> storage_proxy::wait_for_hints_to_be_replayed(std::vector<gms::inet_address> endpoints, seastar::abort_source& as) {
+    auto& db = _db.local();
+    const auto timeout_in_ms = db.get_config().wait_for_hint_replay_before_repair_in_ms();
+    if (timeout_in_ms <= 0) {
+        slogger.debug("wait_for_hint_replay_before_repair_in_ms is zero, so I'm not waiting for other nodes to replay their hints");
+        co_return;
+    }
+
+    const auto deadline = lowres_clock::now() + std::chrono::milliseconds(timeout_in_ms);
+
+    slogger.debug("Coordinating a request to wait until all hints are sent on {} nodes", endpoints.size());
+
+    co_await parallel_for_each(endpoints, [this, &endpoints_ = endpoints, deadline_ = deadline, &as] (gms::inet_address addr) -> future<> {
+        const auto deadline = deadline_;
+        const auto& endpoints = endpoints_;
+
+        const std::chrono::seconds retry_interval{10};
+
+        // Step 1: Create a hint queue sync point
+        utils::UUID sync_point;
+        try {
+            sync_point = co_await _messaging.send_hint_sync_point_create({ addr, 0 }, lowres_clock::now() + retry_interval, endpoints, deadline);
+        } catch (...) {
+            slogger.debug("Failed to create a sync point for {}. I won't wait for hints to be replayed on this node");
+            throw;
+        }
+
+        // Step 2: Wait until hints are replayed
+        // Perform queries every 10 seconds
+        bool reached_sync_point = false;
+        auto next_try_time = lowres_clock::now();
+        while (!reached_sync_point && !as.abort_requested()) {
+            auto now = lowres_clock::now();
+
+            // Wait until the next try time or the deadline, whichever is first
+            const auto wait_until = std::min(next_try_time, deadline);
+            if (wait_until > now) {
+                co_await sleep_abortable(wait_until - now, as);
```

Comment:
The check interval is quite large which means we can waste time sleeping while the check is ok now.

I think we can simplify the logic a lot.

Always sleep 1 second and check in a loop.

If the time is larger than the deadline, stop the loop.

The timeout passing the send_hint_sync_point_check can be always be dealine, in normal cases the  send_hint_sync_point_check should return immediately.

Reply:
Looking at the handler, since we already have a meta data for the hint sync,

sp._hint_queue_checkpoints

Alternatively, we can make storage_proxy::check_hint_queue_sync_point return until either sync is done or sync is timeout or sync is aborted. 

On abort on the sender side, we send a abort cmd to the receiver. 

This makes the code a lit more complicated though. 

I guess the 1 second per node query approach is good enough. We can tolerate the overhead and the max over-sleeping time of 1 second per node.

Reply:
> Alternatively, we can make storage_proxy::check_hint_queue_sync_point return until either sync is done or sync is timeout or sync is aborted.

My previous approach (v1) did exactly this. The repair coordinator would send `send_hint_sync_point_check` request with the timeout equal to the hint replay timeout. It would retry it in case of temporary errors. However, this caused follower nodes to hang (probably waiting for the timeout) when trying to shut them down, so I remade it and now it periodically sends requests which finish quickly.

Reply:
> The check interval is quite large which means we can waste time sleeping while the check is ok now.
>
> I think we can simplify the logic a lot.
>
> Always sleep 1 second and check in a loop.
>
> If the time is larger than the deadline, stop the loop.
>
> The timeout passing the send_hint_sync_point_check can be always be dealine, in normal cases the send_hint_sync_point_check should return immediately.

Done

Reply:
It is much easier to read the code  now.